### PR TITLE
Add support to send IM events

### PR DESCRIPTION
**Pull request by yunhanw-google**
Summary of Changes:

-- When receiving read requested, store interested event path inside
clusterInfo as eventClusterInfoList, reporting engine would fetch
events via EventManamgenment from circular event buffers according to
eventClusterInfoList, and finally send IM report via reader
handler.
-- Reconstruct ClusterInfo so that it can represent the interested
attribute and event efficiently.

### CODE REVIEW

File: src/app/ClusterInfo.h [hunk 1]
```diff
@@ -25,17 +25,39 @@ namespace chip {
...
+    NodeId mNodeId         = 0;
+    ClusterId mClusterId   = 0;
+    ListIndex mListIndex   = 0;
+    FieldId mFieldId       = 0;
+    EndpointId mEndpointId = 0;
+    bool mDirty            = false;
+    Type mType             = Type::kInvalid;
+    ClusterInfo * mpNext   = nullptr;
+    EventId mEventId       = 0;
+    /* For better structure alignment
```
Comment (andy31415) [inline src/app/ClusterInfo.h:49]: This comment may not be as clear. Maybe state as:

above ordering is by bit-size to ensure least amount of memory alignment padding. Changing order to something more natural (e.g. clusterid before nodeid) will result in extra memory alignment padding.

File: src/app/ReadHandler.cpp [hunk 2]
```diff
@@ -231,5 +271,47 @@ void ReadHandler::MoveToState(const HandlerState aTargetState)
...
+    {
+        if ((mLastScheduledEventNumber[i] != 0) && mSelfProcessedEvents[i] <= mLastScheduledEventNumber[i])
+        {
+            mCurrentPriority = static_cast<PriorityLevel>(i);
+            return;
+        }
+    }
+
+    mCurrentPriority = PriorityLevel::Invalid;
+    return;
```
Comment (andy31415) [inline src/app/ReadHandler.cpp:314]: nit: remove extra return - not needed.

File: src/app/EventManagement.cpp [hunk 3]
```diff
@@ -787,6 +789,27 @@ CHIP_ERROR EventManagement::ScheduleFlushIfNeeded(EventOptions::Type aUrgent)
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR EventManagement::SetScheduledEventEndpoint(EventNumber * apEventEndpoints)
+{
+    CHIP_ERROR err                    = CHIP_NO_ERROR;
```
Comment (andy31415) [inline src/app/EventManagement.cpp:794]: err seems never set. Should we just end this with CHIP_NO_ERROR?

is there a possibility of error here?

File: src/app/EventManagement.h [hunk 4]
```diff
@@ -449,6 +451,19 @@ class EventManagement
...
 
+    /**
+     * @brief
+     *   IsValid returns whether the EventManagement instance is valid
+     */
+    bool IsValid(void) { return EventManagementStates::Shutdown != mState; };
+
+    /**
+     *  Logger would save last logged event number for each logger buffer into schedule event number array
+     *
```
Comment (andy31415) [inline src/app/EventManagement.h:462]: nit: remove some of the empty whitespace here.

**Review by andy31415 — APPROVED** [review]

File: src/app/ReadHandler.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp]: Done, thanks.

File: src/app/EventManagement.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/EventManagement.cpp]: never error, remove the error, thanks

File: src/app/EventManagement.h (hunk 4 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h]: Done, thanks

File: src/app/ReadHandler.cpp [hunk 5]
```diff
@@ -231,5 +271,47 @@ void ReadHandler::MoveToState(const HandlerState aTargetState)
     ChipLogDetail(DataManagement, "IM RH moving to [%s]", GetStateStr());
 }
 
+bool ReadHandler::CheckEventClean(EventManagement & aEventManager)
+{
+    if (mCurrentPriority == PriorityLevel::Invalid)
+    {
+        // Upload is not in middle, previous mLastScheduledEventNumber is not valid, Check for new events, and set a checkpoint
+        for (uint8_t i = 0; i < ArraySize(mSelfProcessedEvents); i++)
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/ReadHandler.cpp]: Done，thanks

File: src/app/ClusterInfo.h (hunk 1 above)
Comment (yunhanw-google) [reply src/app/ClusterInfo.h:49]: Thanks, done, agree.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Comment by woody-apple** [comment]
@vivien-apple @bzbarsky-apple ?

**Pull request merged**